### PR TITLE
Macros for methods with Metadata

### LINKED ISSUE #20: Metadata support for Macros

### PR DESCRIPTION
**Pull request by tomusdrw**
Closes #20

### CODE REVIEW

File: core/src/io.rs [hunk 1]
```diff
@@ -157,32 +149,40 @@ impl<T: Metadata> MetaIoHandler<T> {
 
 /// Simplified `IoHandler` with no `Metadata` associated with each request.
 #[derive(Default)]
-pub struct IoHandler(MetaIoHandler<()>);
+pub struct IoHandler<M: Metadata = ()>(MetaIoHandler<M>);
```
Comment (debris) [inline core/src/io.rs]: why did you change that?

File: macros/src/auto_args.rs [hunk 2]
```diff
@@ -101,6 +147,11 @@ pub trait WrapAsync<B: Send + Sync + 'static> {
 	fn wrap_rpc(&self, base: &B, params: Params) -> BoxFuture<Value, Error>;
 }
 
+/// Wrapper trait for meta RPC functions.
+pub trait WrapMeta<B: Send + Sync + 'static, M: Metadata> {
+	fn wrap_rpc(&self, base: &B, params: Params, meta: M) -> BoxFuture<Value, Error>;
```
Comment (rphmeier) [inline macros/src/auto_args.rs]: limitation here is that all metadata-taking methods must be async?

File: core/src/io.rs (hunk 1 above)
Comment (tomusdrw) [inline core/src/io.rs]: To support writing test cases not using metadata easier:

```rust
assert_eq!(io.handle_request_sync(&request), Some(response));
```

and for meta you need to do:
```rust
assert_eq!((*io).handle_request_sync(&request, meta), Some(response));
```
or just use `MetaIoHandler` directly.

File: macros/src/auto_args.rs (hunk 2 above)
Comment (tomusdrw) [inline macros/src/auto_args.rs]: Yes. Do you think we should support sync meta too?

Comment (rphmeier) [inline macros/src/auto_args.rs]: IMO it should all be async now that futures are integrated. But I suppose this limitation is fine.
